### PR TITLE
all: add support for migrate down

### PR DESCRIPTION
**Pull request by masseelch**
Have a look here: https://github.com/ariga/atlas-action/actions/runs/8423912059/job/23066659095?pr=141#step:8:26

### CODE REVIEW

File: atlasaction/action.go [hunk 1]
```diff
@@ -79,6 +79,107 @@ func MigrateApply(ctx context.Context, client *atlasexec.Client, act *githubacti
...
+		ToVersion: act.GetInput("to-version"),
+		ToTag:     act.GetInput("to-tag"),
+		Amount:    a,
+		Context: &atlasexec.DeployRunContext{
+			TriggerType:    atlasexec.TriggerTypeGithubAction,
+			TriggerVersion: Version,
+		},
+		Vars: vars,
+	}
+	if a := act.GetInput("amount"); a != "" {
```
Comment (ronenlu) [inline atlasaction/action.go:119]: Why you do it twice?

File: README.md [hunk 2]
```diff
@@ -256,6 +257,38 @@ All inputs are optional as they may be specified in the Atlas configuration file
...
+* `url` - The URL of the target database.  For example: `mysql://root:pass@localhost:3306/dev`.
+* `dir` - The URL of the migration directory to apply.  For example: `atlas://dir-name` for cloud
+   based directories or `file://migrations` for local ones.
+* `config` - The URL of the Atlas configuration file.  By default, Atlas will look for a file
+  named `atlas.hcl` in the current directory. For example, `file://config/atlas.hcl`.
+  Learn more about [Atlas configuration files](https://atlasgo.io/atlas-schema/projects). 
+* `env` - The environment to use from the Atlas configuration file.  For example, `dev`.
+* `amount` - The amount of applied migrations to revert, defaults to 1.
+* `to-version` - To which version to revert.
+* `to-version` - To which tag to revert.
```
Comment (a8m) [inline README.md:277]:
```suggestion
* `to-tag` - To which tag to revert.
```

File: README.md [hunk 3]
```diff
@@ -256,6 +257,38 @@ All inputs are optional as they may be specified in the Atlas configuration file
...
+* `wait-interval` - Time in seconds between different migrate down attempts, useful when waiting for plan approval. 
+* `wait-timeout` - Time after which no other retry attempt is made and the action exits.
```
Comment (a8m) [inline README.md:279]:
```suggestion
* `wait-timeout` - Time after which no other retry attempt is made and the action exits.
* `wait-interval` - Time in seconds between retry attempts, useful when waiting for plan approval. Defaults to 1s.
```
Keeps it a default maybe.

Comment (a8m) [reply README.md:279]: It made sense to me to change the order

File: atlasaction/action.go [hunk 4]
```diff
@@ -79,6 +79,107 @@ func MigrateApply(ctx context.Context, client *atlasexec.Client, act *githubacti
...
+		wii, err := strconv.Atoi(wi)
+		if err != nil {
+			return fmt.Errorf(`parsing "wait-interval": %w`, err)
+		}
+		wti, err := strconv.Atoi(wt)
+		if err != nil {
+			return fmt.Errorf(`parsing "wait-timeout": %w`, err)
+		}
+		interval = time.Duration(wii) * time.Second
+		timeout = time.Duration(wti) * time.Second
```
Comment (a8m) [inline atlasaction/action.go:143]: Maybe accept a duration value. It's more flexible and that's what used by (some of) Atlas flags

File: atlasaction/action.go (hunk 1 above)
Comment (masseelch) [reply atlasaction/action.go:119]: Good question. I guess it's an artifact of a refacto. Good catch

File: atlasaction/action.go [hunk 5]
```diff
@@ -79,6 +79,107 @@ func MigrateApply(ctx context.Context, client *atlasexec.Client, act *githubacti
...
+			act.SetOutput("error", err.Error())
+			return err
+		}
+		if run.Error != "" {
+			act.SetOutput("error", run.Error)
+			return errors.New(run.Error)
+		}
+		// Break the loop if no wait / retry is configured.
+		if run.Status != StatePending || timeout == 0 || time.Since(started) >= timeout {
+			act.Warningf("plan has not been approved in configured waiting period, exiting")
```
Comment (a8m) [inline atlasaction/action.go:160]: Is this message necessary? Approval flow is optional, and might be applied immediately.

Comment (masseelch) [reply atlasaction/action.go:160]: In that case this condition is not met and nothin it printed. If you use it in an example it makes sense. See my link in the PR description.

**Review by a8m — COMMENTED** [review]
Excellent

File: README.md (hunk 3 above)
Comment (a8m) [reply README.md:279]: I meant adding default only to the interval one. Default to 1s or 5s.

File: atlasaction/action_test.go [hunk 6]
```diff
@@ -119,6 +120,141 @@ func TestMigrateApply(t *testing.T) {
...
+		tt.setInput("env", "test")
+
+		// Defaults have been tested before by other tests.
+
+		tt.setInput("wait-interval", "1")
+		err := MigrateDown(context.Background(), tt.cli, tt.act)
+		require.EqualError(t, err, "both \"wait-interval\" and \"wait-timeout\" must be given or be empty, got \"1\" and \"\"")
+		tt.setInput("wait-interval", "")
+
+		tt.setInput("wait-timeout", "1")
```
Comment (ronenlu) [inline atlasaction/action_test.go:237]:
```suggestion

		tt.setInput("wait-interval", "")
		tt.setInput("wait-timeout", "1")
```

File: migrate/down/action.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    description: The amount of applied migrations to revert. Mutually exclusive with `to-tag` and `to-version`.
+    required: false
+  to-version:
+    description: The version to revert to. Mutually exclusive with `amount` and `to-tag`.
+    required: false
+  to-tag:
+    description: The tag to revert to. Mutually exclusive with `amount` and `to-version`.
+    required: false
+  working-directory:
+    description: Atlas working directory, default is project root
```
Comment (ronenlu) [inline migrate/down/action.yml:39]:
```suggestion
    description: Atlas working directory, default is project root.
```

File: atlasaction/action.go [hunk 8]
```diff
@@ -79,6 +79,99 @@ func MigrateApply(ctx context.Context, client *atlasexec.Client, act *githubacti
...
+			return errors.New(run.Error)
+		}
+		// Break the loop if no wait / retry is configured.
+		if run.Status != StatePending || timeout == 0 || time.Since(started) >= timeout {
+			if timeout != 0 {
+				act.Warningf("plan has not been approved in configured waiting period, exiting")
+			}
+			break
+		}
+		act.Infof("plan approval pending, review here: %s", run.URL)
```
Comment (a8m) [inline atlasaction/action.go:155]: Print it only once (in the first iteration).

**Review by a8m — APPROVED** [review]

**Pull request merged**